vcservice/vcservice.go

```diff
@@ -0,0 +1,278 @@
+			if err != nil {
+				c.log.Error(err)
+				stream = nil
+				time.Sleep(ReconnectSleep)
+				continue
+			}
+		}
+
+		response, err := stream.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// Nice reconnect messages
+				c.log.Errorf("vc-service server is unavailable, retrying in %s...", ReconnectSleep.String())
+			} else {
+				c.log.Errorf("Error receiving message: %s", err)
+			}
+
+			// Stream is no longer useful. Need to get a new one on reconnect
+			stream = nil
+
+			// Attempt reconnect. On the next loop iteration, stream == nil check will be hit, and assuming we've
+			// reconnected at that point, a new stream will be opened and authorized
+			c.reconnect()
+
+			time.Sleep(ReconnectSleep)
+			continue
+		}
+
+		// Handle any backend work that needs to be done
+		if err := c.handleResponse(response); err != nil {
+			c.log.Errorf("Unable to handle VCEvent: %s", err)
+		}
+
+		// Ship message to frontend
+		c.AttachedStreamsMutex.RLock()
+		for _, as := range c.AttachedStreams {
+			as.EventsCh <- response
+		}
+		c.AttachedStreamsMutex.RUnlock()
+	}
+}
+
+// authorize opens a GRPC connection to vc-service. It is called by vcservice.Client.Start()
+func (c *Client) authorize() (protos.VCService_ConnectClient, error) {
+	authRequest := &protos.ConnectAuthRequest{
+		ApiToken: c.PersistentConfig.VCServiceToken,
+	}
+
+	return c.Client.Connect(context.Background(), authRequest)
+}
+
+// handleResponse receives a dynamic replay message and determines which method should handle it based on the payload
+func (c *Client) handleResponse(resp *protos.VCEvent) error {
+	switch resp.Type {
+	case protos.VCEvent_AUTH_RESPONSE:
+		return c.handleAuthResponse(resp)
+	case protos.VCEvent_GITHUB:
+		return c.handleGithubEvent(resp)
+	case protos.VCEvent_GITLAB:
+		return errors.New("VCEvent_GITLAB is not implemented yet")
+	case protos.VCEvent_BITBUCKET:
+		return errors.New("VCEvent_BITBUCKET is not implemented yet")
+	}
+
+	return fmt.Errorf("unknown VCEvent type '%s'", resp.Type)
+}
+
+// handleAuthResponse handles a AUTH_RESPONSE payload from a VCServiceServer protobuf message
+func (c *Client) handleAuthResponse(event *protos.VCEvent) error {
+	authResponse := event.GetAuthResponse()
+	if authResponse == nil {
+		if err := c.Conn.Close(); err != nil {
+			c.log.Error("could not cleanly disconnect from server")
+		}
+		return errors.New("Received invalid authentication response from server")
+	}
+
+	if authResponse.Authorized == false {
+		if err := c.Conn.Close(); err != nil {
+			c.log.Error("could not cleanly disconnect from server")
+		}
+
+		err := fmt.Errorf("could not authenticate: %s", authResponse.Message)
+		c.log.Error(err)
+		return err
+	}
+
+	c.log.Info("Connection authorized. Waiting for vc events...")
+
+	return nil
+}
+
+func (c *Client) handleGithubEvent(event *protos.VCEvent) error {
+	payload := event.GetGithubEvent()
+	switch payload.GetType() {
+	case protos.GithubEvent_INSTALL_CREATED:
+		// TODO
```

Comment:
Is there a need to spell each one of these out? Can we just iterate over a map verifying if it's a "known github event"?

Reply:
They are separate as we will eventually need to act on the messages somehow instead of just shipping them to the frontend